src/graph/SetExecutor.cpp

```diff
@@ -0,0 +1,384 @@
+                break;
+            case SetSentence::Operator::INTERSECT:
+                doIntersect();
+                break;
+            case SetSentence::Operator::MINUS:
+                doMinus();
+                break;
+            default:
+                LOG(ERROR) << "Unknown operator: " << sentence_->op();
+        }
+    };
+    folly::collectAll(futures_).via(runner).thenValue(cb);
+}
+
+void SetExecutor::doUnion() {
+    VLOG(3) << "Do Union.";
+    if (leftResult_ == nullptr) {
+        VLOG(3) << "Union has right result.";
+        finishExecution(std::move(rightResult_));
+        return;
+    }
+
+    if (rightResult_ == nullptr) {
+        VLOG(3) << "Union has left result.";
+        finishExecution(std::move(leftResult_));
+        return;
+    }
+
+    Status status = checkSchema();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(status);
+        return;
+    }
+
+    auto leftRows = leftResult_->getRows();
+    auto rightRows = rightResult_->getRows();
+    if (!castingMap_.empty()) {
+        auto stat = doCasting(rightRows);
+        if (!stat.ok()) {
+            DCHECK(onError_);
+            onError_(status);
+            return;
+        }
+    }
+
+    std::vector<cpp2::RowValue> rows;
+    if (sentence_->distinct()) {
+        rows = doDistinct(leftRows, rightRows);
+    } else {
+        leftRows.insert(leftRows.end(), rightRows.begin(), rightRows.end());
+        rows = std::move(leftRows);
+    }
+
+    finishExecution(std::move(rows));
+    return;
+}
+
+Status SetExecutor::checkSchema() {
+    auto leftSchema = leftResult_->schema();
+    auto rightSchema = rightResult_->schema();
+    auto leftIter = leftSchema->begin();
+    auto rightIter = rightSchema->begin();
+
+    auto index = 0u;
+    while (leftIter && rightIter) {
+        auto *colName = rightIter->getName();
+        if (leftIter->getType() != rightIter->getType()) {
+            castingMap_.emplace(index, leftIter->getType());
+        }
+
+        colNames_.emplace_back(std::string(colName));
+
+        ++index;
+        ++leftIter;
+        ++rightIter;
+    }
+
+    if (leftIter || rightIter) {
+        return Status::Error("Field count not match.");
+    }
+
+    resultSchema_ = std::move(leftSchema);
+
+    return Status::OK();
+}
+
+Status SetExecutor::doCasting(std::vector<cpp2::RowValue> &rows) const {
+    for (auto &row : rows) {
+       auto cols = row.get_columns();
+       for (auto &pair : castingMap_) {
+           auto stat =
+               InterimResult::castTo(&cols[pair.first], pair.second.get_type());
+            if (!stat.ok()) {
+                return stat;
+            }
+       }
+       row.set_columns(std::move(cols));
+    }
+
+    return Status::OK();
+}
+
+
+std::vector<cpp2::RowValue> SetExecutor::doDistinct(
+        std::vector<cpp2::RowValue> &leftRows,
+        std::vector<cpp2::RowValue> &rightRows) const {
+    std::vector<cpp2::RowValue> rows;
+    for (auto &lr : leftRows) {
```

Comment:
It would be better to do dinstict by std::sort and std::unique.

Reply:
If the dataset was large or we can't directly modify leftRows/rightrows, it might be better and intuitive to use unordered_set  to do distinct.

Reply:
> If the dataset was large or we can't directly modify leftRows/rightrows, it might be better and intuitive to use unordered_set to do distinct.

The row value is an object that contains vector, I wonder if it would perform well when we provide hash for vector.

Reply:
> It would be better to do dinstict by std::sort and std::unique.

This seems do better performance, for reducing the occurrence of resize and time complexity, I buy in this idea.